File: Leetcode_with_py/79exist.py
class Solution(object):
    directions = [(0, -1), (-1, 0), (0, 1), (1, 0)]
    
    def exist(self, board, word):
        m = len(board)
        if m == 0:
            return False
        n = len(board[0])
        marked = [[False for _ in range(n)] for _ in range(m)]
        for i in range(m):
            for j in range(n):
                # 对每一个格子都从头开始搜索
                if self.searchWord(board, word, 0, i, j, marked, m, n):
                    return True
        return False

    def searchWord(self, board, word, index, x, y, marked, m, n):
        # 先写递归终止条件
        if index == len(word) - 1:
            return board[x][y] == word[index]
        # 中间匹配了，再继续搜索
        if board[x][y] == word[index]:
            # 先占住这个位置，搜索不成功的话，要释放掉
            marked[x][y] = True
            for direction in self.directions:
                newX = x + direction[0]
                newY = y + direction[1]
                # 注意：如果这一次 search word 成功的话，就返回
                if 0 <= newX < m and 0 <= newY < n and not marked[newX][newY] and \
                        self.searchWord(board, word, index + 1, newX, newY, marked, m, n):
                    return True
            marked[x][y] = False
        return False

File: Leetcode_with_Cpp/79exist.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if (board.empty()) return false;
        int m = board.size(), n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        bool res = false;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                backtrack(board, word, i, j, 0, visited, res);
            }
        }
        return res;
    }
    void backtrack(vector<vector<char>>& board, string & word, int i, int j, int pos, vector<vector<bool>>& visited, bool & res) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size())
            return;
        if (visited[i][j] || res || board[i][j] != word[pos])
            return;
        if (pos == word.size() - 1) {
            res = true;
            return;
        }
        visited[i][j] = true;
        backtrack(board, word, i+1, j, pos+1, visited, res);
        backtrack(board, word, i-1, j, pos+1, visited, res);
        backtrack(board, word, i, j+1, pos+1, visited, res);
        backtrack(board, word, i, j-1, pos+1, visited, res);
        visited[i][j] = false;
    }
};